File: packages/flutter/lib/src/resources/mcps.dart
/// MCP servers resource — `/api/mcps`.
library;

import 'base.dart';

class McpsResource extends BaseResource {
  const McpsResource(super.client);

  Future<List<Map<String, dynamic>>> list() async {
    final data = await client.request('GET', 'mcps');
    if (data is List) return data.whereType<Map<String, dynamic>>().toList();
    return const [];
  }

  Future<Map<String, dynamic>> get(String id) async {
    final data = await client.request('GET', 'mcps/$id');
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> create(Map<String, dynamic> payload) async {
    final data = await client.request('POST', 'mcps', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> update(
      String id, Map<String, dynamic> payload) async {
    final data = await client.request('PUT', 'mcps/$id', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<void> delete(String id) => client.request('DELETE', 'mcps/$id');

  /// `POST /api/mcps/{id}/test` — test the MCP server connection.
  Future<Map<String, dynamic>> test(String id) async {
    final data = await client.request('POST', 'mcps/$id/test');
    return (data as Map<String, dynamic>?) ?? {};
  }

  /// `GET /api/mcps/{id}/tools` — list available tools.
  Future<Map<String, dynamic>> tools(String id) async {
    final data = await client.request('GET', 'mcps/$id/tools');
    return (data as Map<String, dynamic>?) ?? {};
  }
}

File: packages/python/src/tbd_agents/resources/mcps.py
from __future__ import annotations

from typing import Any, Mapping

from tbd_agents.types import AllowedToolsUpdate, McpServer, McpServerCreate, McpServerUpdate, McpTestResult, McpTools

from .base import BaseResource


class McpsResource(BaseResource):
    def create(self, payload: McpServerCreate | Mapping[str, Any]) -> McpServer:
        return self._post("/mcps", payload, model=McpServer)

    def list(self) -> list[McpServer]:
        return self._get_list("/mcps", model=McpServer)

    def get(self, server_id: str) -> McpServer:
        return self._get(f"/mcps/{server_id}", model=McpServer)

    def update(self, server_id: str, payload: McpServerUpdate | Mapping[str, Any]) -> McpServer:
        return self._put(f"/mcps/{server_id}", payload, model=McpServer)

    def delete(self, server_id: str) -> None:
        self._delete(f"/mcps/{server_id}")

    def test(self, server_id: str) -> McpTestResult:
        return self._post(f"/mcps/{server_id}/test", model=McpTestResult)

    def list_tools(self, server_id: str) -> McpTools:
        return self._get(f"/mcps/{server_id}/tools", model=McpTools)

    def update_tools(
        self,
        server_id: str,
        allowed_tools: list[str] | AllowedToolsUpdate,
    ) -> AllowedToolsUpdate:
        payload = allowed_tools
        if isinstance(allowed_tools, list):
            payload = AllowedToolsUpdate(allowed_tools=allowed_tools)
        return self._put(f"/mcps/{server_id}/tools", payload, model=AllowedToolsUpdate)

